src/import-data/ImportData.ts

```diff
@@ -75,7 +75,7 @@ export const ImportData = (props: ImportDataProps): ReactNode => {
         jobId,
         wdsProxyUrl: wdsUrl,
       };
-      asyncImportJobStore.update(Utils.append(newJob));
+      asyncImportJobStore.update(Utils.append<AsyncImportJob>(newJob));
```

Comment:
The old append function was not type-safe. The new function assumes that the array that you will append to is a list of the same time as the item that you are appending. That works in most cases, but asyncImportJobStore is a union type of `AzureAsyncImportJob | GCPAsyncImportJob`, so you will get an error unless you cast this to the union type. 

As an alternative, I believe you could cast newJob to AzureAsyncImportJob.

Reply:
Not sure why casting would make a difference as `newJob` is already type `AzureAsyncImportJob`. I'm not sure why the new `append` can't infer the type correctly from context here, but adding a cast or a type argument doesn't seem like a great solution.

> That works in most cases, but asyncImportJobStore is a union type of AzureAsyncImportJob | GCPAsyncImportJob, so you will get an error unless you cast this to the union type.

I am not a typescript expert, but I thought the point of a union type was that it could be either one of the types. Since `AzureAsyncImportJob` is one of the two types in the union type, and since `newJob` is that type, why is the type argument needed?

In any event it looks like this form is common in our code, and if there's no way to fix this without adding a cast or type argument, I would create a helper function that both appends and use that instead of calling `asyncImportJobStore` directly.

Reply:
I'm guessing that `append` was inferring its type parameter from it's argument, which is an `AzureAsyncImportJob` and then returning a function that takes `AzureAsyncImportJob[]`. Since `asyncImportJobStore.update` calls the function with a `AsyncImportJob[]` that may include `GCPAsyncImportJob`s, TypeScript says that it's not valid.